File: backend/src/InsuraTech.Domain/Policies/ValueObjects/PolicyNumber.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using InsuraTech.Domain.Common;

namespace InsuraTech.Domain.Policies.ValueObjects
{
    public sealed class PolicyNumber : ValueObject
    {
        public string Value {  get; }
        private PolicyNumber(string value) => Value = value;

        public static PolicyNumber Create(int year, long sequence)
        {
            if (year < 2000 || year > 9999)
                throw new ArgumentOutOfRangeException("Year is out of valid range", nameof(year));
            if (sequence <= 0)
                throw new ArgumentException("Sequence must be greater than zero.",nameof(sequence));

            var value = $"POL-{year}-{sequence:D8}";
            return new PolicyNumber(value);
        }
        public static PolicyNumber Parse(string value)
        {
            if(string.IsNullOrWhiteSpace(value))
                throw new ArgumentNullException("Policy number cannot be empty. ", nameof(value));

            var parts = value.Split('-');
            if(parts.Length != 3 || parts[0] != "POL" )
                throw new ArgumentException($"Invalid policy number format: '{value}'. ", value);
            return new PolicyNumber(value);
        }



        protected override IEnumerable<object> GetEqualityComponets()
        {
            yield return Value;
        }
        public override string ToString() => Value;
    }
}

File: backend_v2/src/main/java/com/insuratech/domain/policies/vo/PolicyNumber.java
package com.insuratech.domain.policies.vo;

import com.insuratech.domain.common.ValueObject;
import com.insuratech.domain.exceptions.InvalidPolicyNumberException;

import java.util.regex.Pattern;

public final class PolicyNumber extends ValueObject {

    private static final Pattern PATTERN = Pattern.compile("^POL-\\d{4}-\\d{6}$");

    private final String value;

    private PolicyNumber(String value) {
        this.value = value;
    }

    public static PolicyNumber of(String value) {
        if (value == null || !PATTERN.matcher(value).matches()) {
            throw new InvalidPolicyNumberException(value);
        }
        return new PolicyNumber(value);
    }

    public static PolicyNumber generate(int year, long sequence) {
        return new PolicyNumber(String.format("POL-%d-%06d", year, sequence));
    }

    public String getValue() {
        return value;
    }

    @Override
    protected Object[] getEqualityComponents() {
        return new Object[]{value};
    }

    @Override
    public String toString() {
        return value;
    }
}
